chia/wallet/wallet_node.py

```diff
@@ -646,9 +646,11 @@ def is_new_state_update(cs: CoinState) -> bool:
         # We only process new state updates to avoid slow reprocessing. We set the sync height after adding
         # Things, so we don't have to reprocess these later. There can be many things in ph_update_res.
         already_checked_ph: Set[bytes32] = set()
-        continue_while: bool = True
-        all_puzzle_hashes: List[bytes32] = await self.get_puzzle_hashes_to_subscribe()
-        while continue_while:
+        while not self._shut_down:
+            await self.wallet_state_manager.create_more_puzzle_hashes()
+            all_puzzle_hashes = await self.get_puzzle_hashes_to_subscribe()
+            if all(puzzle_hash in already_checked_ph for puzzle_hash in all_puzzle_hashes):
+                break
```

Comment:
although a simplification, this code is still not very straight-forward (I don't think). It would seem more natural to compute the set subtraction `all_puzzle_hashes - already_checked_ph` here, and if it's an empty set we're done, otherwise we iterate over them to subscribe. Now we chunk all puzzle hashes, just to still filter out the ones in `already_checked_ph`, so chunks could become empty in some cases. (filtering before chunking would seem to make more sense).

However, since we keep updating the `already_checked_ph` at the end of the loop, it looks like all this juggling of *all* puzzle hashes and `already_checked_ph` is just a roundabout way to get what we really want: the newly created puzzle hashes.

Would it be crazy to make `create_more_puzzle_hashes()` to just return the new ones? Then (I believe) we could just call `get_puzzle_hashes_to_subscribe()` once, outside the loop, and only consider newly created ones inside the loop.

Reply:
> although a simplification, this code is still not very straight-forward (I don't think). It would seem more natural to compute the set subtraction `all_puzzle_hashes - already_checked_ph` here, and if it's an empty set we're done, otherwise we iterate over them to subscribe. Now we chunk all puzzle hashes, just to still filter out the ones in `already_checked_ph`, so chunks could become empty in some cases. (filtering before chunking would seem to make more sense).

Agree, makes sense, see [31c2ca9](https://github.com/Chia-Network/chia-blockchain/pull/14185/commits/31c2ca9580a863b12e2e8e6dedf7006cbd087be1).

> Would it be crazy to make `create_more_puzzle_hashes()` to just return the new ones? Then (I believe) we could just call `get_puzzle_hashes_to_subscribe()` once, outside the loop, and only consider newly created ones inside the loop.

Might make sense but rather not here in this PR?